### PR TITLE
Clean up AllocRef implementation and documentation

### PR DESCRIPTION
**Pull request by TimDiekmann**
r? @Amanieu

### CODE REVIEW

File: library/core/src/alloc/mod.rs [hunk 1]
```diff
@@ -192,12 +197,9 @@ pub unsafe trait AllocRef {
             "`new_size` must be greater than or equal to `layout.size()`"
         );
 
-        let new_layout =
-            // SAFETY: the caller must ensure that the `new_size` does not overflow.
-            // `layout.align()` comes from a `Layout` and is thus guaranteed to be valid for a Layout.
-            // The caller must ensure that `new_size` is greater than or equal to zero. If it's equal
-            // to zero, it's catched beforehand.
```
Comment (TimDiekmann) [inline library/core/src/alloc/mod.rs:199]: No, the caller don't have to, `self.alloc` is well defined for zero-sized layouts. Thus I removed this. This does not hold anyway :D

File: library/std/src/alloc.rs [hunk 2]
```diff
@@ -131,40 +131,82 @@ pub use alloc_crate::alloc::*;
...
+                    GlobalAlloc::alloc(self, layout)
+                };
+                let ptr = NonNull::new(raw_ptr).ok_or(AllocErr)?;
+                Ok(NonNull::slice_from_raw_parts(ptr, size))
+            },
+        }
+    }
+
+    #[inline]
+    fn grow_impl(
```
Comment (Amanieu) [inline library/std/src/alloc.rs:153]: Same here.

**Comment by Amanieu** [comment]
Do you think this is likely to have a perf impact? Let me know so I can set rollup=never.

**Comment by TimDiekmann** [comment]
Hmm, I don't think so, but you can rollup=never just to be sure. If there is one, it shouldn't be more than 0.1% at most.

**Pull request merged**